zone.go

```diff
@@ -115,6 +120,79 @@ func (z *Zone) IsInRootZone() bool {
 	return z.IsTLD() && z.IsDelegated()
 }
 
+// Checks to see if a domain is valid according to character set restrictsion
+// on the zone
+func (z *Zone) IsValidDomain(domain string) bool {
+	suffix := "." + z.Domain
+	if !strings.HasSuffix(domain, suffix) {
+		return false
+	}
+	if len(z.CodePoints) == 0 {
+		return true
+	}
+	// Don't check the suffix since we've already done that
+	return domainInPoints(domain[:len(domain)-len(suffix)], z.CodePoints)
+}
+
+// Return the IDN table language the domain matches. Returns an empty string
+// if no IDN tables are defined for the zone or if the domain is not an IDN.
+func (z *Zone) IDNTable(domain string) (string, error) {
+	suffix := "." + z.Domain
+	if !strings.HasSuffix(domain, suffix) {
+		return "", errors.New("domain is not a member of the zone")
+	}
+
+	// A blank IDN table language indicates no IDN tables are present
+	if len(z.CodePoints) == 0 {
+		return "", nil
+	}
+
+	// Don't check the suffix since we've already done that, and make sure
+	// we are working with the unicode form, since the ASCII form will never
+	// match an IDN table
+	toCheck, _ := idna.ToUnicode(domain[:len(domain)-len(suffix)])
+
+	if !domainInPoints(toCheck, z.CodePoints) {
+		return "", errors.New("domain is not a valid member of the zone")
+	}
+
+	// Ensure it is an IDN
+	if domainInPoints(toCheck, asciiCodePoints) {
+		return "", nil
+	}
+
+	for lang, codePoints := range z.IDNTables {
+		if domainInPoints(toCheck, codePoints) {
+			return lang, nil
+		}
+	}
+
+	return "", errors.New("domain is not a valid IDN of the zone")
+}
+
+var asciiCodePoints = []rune{'-', '-', '0', '9', 'a', 'z'}
+
+func domainInPoints(domain string, codePoints []rune) bool {
+	for _, char := range domain {
+		if !charInPoints(char, codePoints) {
+			return false
+		}
+	}
+	return true
+}
+
+func charInPoints(char rune, codePoints []rune) bool {
```

Comment:
First, rename to `runeInCodePoints`. Rename args `char` -> `c` and `codePoints` -> `points`.

Second, let’s optimize this by recursively splitting `points` and testing against outer bounds.

1. Assuming the code points are sorted, then we can test `c >= points[0]` and `c <= points[len(points)-1]`.
2. Then split `points` around a pivot in the middle, and recursively call `runeInCodePoints` (or iterate, rather than recurse) until we prove or disprove `c` is within the range of code points.

Go has a couple examples of similar code here: https://github.com/golang/go/blob/master/src/unicode/letter.go#L125